src/include/server/mir/shell/token_authority.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_SHELL_TOKEN_AUTHORITY_H_
+#define MIR_SHELL_TOKEN_AUTHORITY_H_
+
+#include "mir/main_loop.h"
+#include "mir/time/alarm.h"
+
+#include <chrono>
+#include <cmath>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unistd.h>
+#include <unordered_set>
+
+namespace mir
+{
+namespace shell
+{
+class TokenAuthority
+{
+public:
+    class Token
+    {
+    public:
+        using RevocationListener = std::function<void(Token const& token)>;
+
+        Token(
+            std::string token,
+            std::unique_ptr<time::Alarm> alarm,
+            std::optional<RevocationListener> revocation_listener) :
+            token{token},
+            alarm{std::move(alarm)},
+            revocation_listener{revocation_listener}
+        {
+            alarm->reschedule_in(timeout_ms);
+        }
+
+        explicit Token(std::string token) :
+            token{token},
+            alarm{nullptr}
+        {
+        }
+
+        operator std::string() const
+        {
+            return token;
+        }
+
+        bool operator==(Token const& token) const
+        {
+            return this->token == token.token;
+        }
+
+        std::size_t hash() const
+        {
+            return std::hash<std::string>{}(token);
+        }
+
+    private:
+        friend TokenAuthority;
+        std::string token;
+        std::shared_ptr<mir::time::Alarm> const alarm;
+        std::optional<RevocationListener> revocation_listener;
+    };
+
+    TokenAuthority(std::shared_ptr<MainLoop> main_loop) :
+        main_loop{main_loop}
+    {
+    }
+
+    auto issue_token(std::optional<Token::RevocationListener> revocation_listener) -> Token
+    {
+        std::scoped_lock lock{mutex};
+
+        auto const generated = generate_token();
+        auto alarm = main_loop->create_alarm(
```

Comment:
This line intermittently returns a null alarm for some reason

Reply:
I  was moving out of an object and into a data member in a constructor. Both had the same name. Then I used the name to call a function. For some reason the compiler chose to use the passed in object instead of `this->object`